contributors/design-proposals/sig-cli/apply_v2.md

```diff
@@ -0,0 +1,91 @@
+It does a couple things:
+
+1. Create / Update / (Delete) the resource based on the resources
+2. Only set / delete fields defined in the file either currently or in during last update from the file
+  - Retain updates to fields made by controllers that were never managed in the file
+3. Merge fields based on the field name, merge lists based on a field contained within the list elements
+
+The apply logic has been difficult to maintain from a contributor perspective, and difficult
+to understand from a user perspective.  Much of this pain has comes from the `apply` command out
+growing its implementation.  In order to continue to support apply and provide better user visibility
+into what it is doing, the command must be re-architected to break client dependencies on the server-side
+implementation, address gaps in diff-patch semantics preventing correctly applying certain fields,
+and to organize the code in a modular and extendable manner.
+
+## Changes
+
+### Use read-update instead of patch
+
+#### Why
+
+Issues with updating the object by calculating and sending a patch to the server
+
+- Patch API is not fully featured enough to represent all of the operations required
+  to update an object and frequently has to be expanded with new side-directives describing
+  how the server should apply the patch.  Updates are done in the form of adding new
+  merge directives to the patch format on the server, and invoking them from the client.
+- Having to update both client and server to fix bugs introduces challenges to supporting
+  version skew, as the client will depend on the server supporting a patch semantic.  There
+  is currently no versioning of the patch semantics, so the client cannot reliably know what
+  patch semantics the server supports.
+- Patch syntax is not human readable, making it hard for humans to evaluate the end state
+  that will result from the patch.  `--dry-run` is nearly useless for understanding the
+  end state that will result from apply, and implementing a diff operation has been exceedingly
+  copmlex.
+- Generating a patch to make a change is more complicated than simply making the change to
+  the object read from the server.  The apply command reads the live object from the server
+  and diffs it against the local configuration.  Building a patch containing special merge
+  directives is a more complicated approach than modifying the object read from the server
+  and updating it with a PUT.
+- Patches provide weaker optimistic locking guarantees than PUT.  When encountering an optimistic
+  lock failure in apply, it is easy to re-read the object and try to update it again.  If the
+  apply command tries to modify a field that was modified since apply read the object, it will fail
+  and optionally be able to show this to a user, whereas this conflict may not be detect with a PATCH
+  operation.
+
+#### New approach
```

Comment:
Why not just move apply to the server via a formal API?

Reply:
@bgrant0607 asked a similar question.

My hope as that the steps in this proposal will make moving the logic to an API easier by reducing the complexity of the existing solution and making it a modular package that does not depend on cobra or other kubectl libraries to diff & update objects.

I'd be open to exposing the re-architected libraries as an API.

Reply:
Because that's how we got strategic merge :P   In the general case, the server _can't_ know whether particular operations should be "replace" or "merge", and trying to capture all that flexibility in the API has created exactly the sort of complexity that motivated this doc.  The alternative is just to do it client-side, where naturally there is more context available to be able to implement the exact semantics intended by the client.

Reply:
Any examples in the API schema where it shouldn't be obvious? In the few ordered lists we have (e.g., command, args), those need to be replaced. All other cases should be merged. In the case of no conflict between the configured state and the live state, a merge and a replacement should be the same.